### ISSUE TITLE
Exception when trying to parse Trans element JSX with interpolation

### ISSUE THREAD

**Issue by doncarron** [i18next/i18next-parser#95]
- Is this a bug or a feature request? bug
- What version are you running (`0.x` or `1.x`)?  1.0.0-beta13
- Which runner (cli, gulp...)? cli

## What I'm trying to do?

Extract translations from a Trans tag that uses interpolation

## How am I doing it?

./node_modules/.bin/i18next 'trans-tmp/components/*.jsx' -c i18n.config.js

Example React code being parsed:

...
                    <DialogHeaderTitle>
                        <Trans>Consumer Information for {{name}}</Trans>
                    </DialogHeaderTitle>
...

Here is my config file:

module.exports = {
    reactNamespace: true,
    locales: ['fr', 'en'],
    output: 'locales'
};

## What was the result?

  i18next Parser
  --------------
  Input:  'trans-tmp/components/*.jsx'
  Output: locales

  [read]  C:\GIT\react-component-final\trans-tmp\components\Component.jsx
  [read]  C:\GIT\react-component-final\trans-tmp\components\ComponentConsumers.jsx
  [read]  C:\GIT\react-component-final\trans-tmp\components\ComponentDependencies.jsx
  [read]  C:\GIT\react-component-final\trans-tmp\components\ComponentList.jsx
  [read]  C:\GIT\react-component-final\trans-tmp\components\Header.jsx
  [read]  C:\GIT\react-component-final\trans-tmp\components\TitleBar.jsx
  [write] C:\GIT\react-component-final\locales\fr\Consumer Information for <1>{{ name.json
  [write] C:\GIT\react-component-final\locales\fr\Consumer Information for <1>{{ name_old.json
  [write] C:\GIT\react-component-final\locales\fr\translation.json
  [write] C:\GIT\react-component-final\locales\fr\translation_old.json
  [write] C:\GIT\react-component-final\locales\en\Consumer Information for <1>{{ name.json
  [write] C:\GIT\react-component-final\locales\en\Consumer Information for <1>{{ name_old.json
  [write] C:\GIT\react-component-final\locales\en\translation.json
  [write] C:\GIT\react-component-final\locales\en\translation_old.json

  Stats:  6 files were parsed
events.js:183
      throw er; // Unhandled 'error' event

Error: ENOENT: no such file or directory, open 'C:\GIT\react-component-final\locales\fr\Consumer Information for <1>{{ name.json'

## What was I expecting?

Proper output translation key: EX:

  "Consumer Information for <1>{{name}}</1>": "Information du consommateur pour <1>{{name}}</1>",

Not sure why the parser is attempting to output the new files with an improperly-parsed key.

**Comment by karellm** [comment]
@MynockSpit Could you take a look at this one by chance?

**Comment by MynockSpit** [comment]
Taking a look.

**Comment by MynockSpit** [comment]
@doncarron 

The namespace is being clobbered by... something. Are you using the translate function to insert translations? i.e. `export default translate('my-translations')(App);` Also, Is that the first string that it fails to translate, or do some succeed before?

**Comment by MynockSpit** [comment]
@karellm 

Potentially related -- I noticed that we're splitting the keys on `.` by default and storing them as nested objects. Meaning, if you happen to have several dots in your keys, you'll end up with a translation file looking something like...

```json
// key: namespace.my.dotted.key
// namespace.json
{  
  "my": {
    "dotted": {
      "key": "value"
    }
  }
}
```

Is that right?

**Comment by doncarron** [comment]
@MynockSpit That's the first string that fails causing the exception. All the others work fine and if I remove the string with the interpolation the process finishes without issue. 

I'm not sure I follow the namespace being clobbered - it works for everything else. It's just the interpolation that seems to be causing a problem. I am not using the translate function yet - just the Trans component.

**Comment by MynockSpit** [comment]
Bizarre. I can't reproduce this -- even with an identical string. 

Some questions: 
- What happens when you add an i18nKey? For example... 
    ```jsx
    <Trans i18nKey="my.cool.key"> ... </Trans>
    ```
- What happens if you add translate to your exponent's export? For example...
    ```js
    import { translate } from "react-i18next";  
    // ...
    export default translate('translations')(MyComponent)
    ```
- Does interpolation work in any other strings?

Alternatively, if you can reproduce this in something like https://stackblitz.com or https://codesandbox.io, that would be amazing. (If not, no worries, I'll keep trying.)

**Comment by doncarron** [comment]
@MynockSpit To answer your first and third questions:

1) Adding a key does cause the extract to succeed, yes. However one of my primary motivators for using i18next was the simple setup for "keyless" (content keyed) translation.

As an aside are React users supposed to change the default interpolation prefix / suffix to just a single curly ({ })? I noted that the double-curlys work for the i18n framework, but not for React's own interpolation.

3) Interpolation has the same problem in any string.

I haven't had a chance to try #2 yet.

**Comment by MynockSpit** [comment]
> As an aside are React users supposed to change the default interpolation prefix / suffix to just a single curly ({ })? I noted that the double-curlys work for the i18n framework, but not for React's own interpolation.

Nah. Single curly is still normal React interpolation. You can use it for i18next-react if you use a key with it. If you're using the content-as-keys, the value will be inserted BEFORE the key is 'generated'.

i.e. 
```jsx
const name = 'Than';
<Trans>My name is {name}.</Trans> // produces a key of `My name is Than.`
<Trans>My name is {{name}}.</Trans> // produces a key of `My name is <0>{{name}}</0>.`
```
 
Yeah, keyless is a nice plus. I have a couple of theories of what might be wrong. In the meantime, try setting `namespaceSeparator` and `keySeparator` to `false` in the i18next-parser config.

**Comment by doncarron** [comment]
@MynockSpit First of all thanks so much for your continued support. I'm really hoping we get this working!

After updating my config as requested the key generates without throwing an exception, however the key generated is wrong. The key produced is:

  "Consumer information for <1>{{name: name}}</1>"

When the debug mode for i18next is looking for:

  "Consumer information for <1>{{name}}</1>"

As an aside I noticed my setup doesn't produce the results you indicated above. What I'm seeing is:

`const name = 'Than';
<Trans>My name is {name}.</Trans> // produces a key of `My name is <1>{name}</1>.`
<Trans>My name is {{name}}.</Trans> // produces a key of `My name is <1>{{name: name}}</1>.``

In the event this information is helpful at all:

    "i18next": "11.3.2",
    "prop-types": "15.6.1",
    "react": "16.3.2",
    "react-dom": "16.3.2",
    "react-i18next": "7.6.1"

**Comment by MynockSpit** [comment]
Woohoo! Progress! That's (likely) it. `{ name }` is ES6 object shorthand for `{ name: name }`. Seems like the transpile to TypeScript is expanding that out.

i18next-parser (and i18next-react and i18next) have a `namespaceSeparator` value that indicates a separator that you use to split along namespace boundaries. We write different namespaces into different files. The default is `:`.

Meaning, a key of `myCoolNamespace:myCoolKey` would produce a file of `myCoolNamespace.json` with a key of `myCoolKey`, and you'd load it under a `myCoolNamespace` key in the i18next config.

I'll dig into it more this evening.

**Comment by MynockSpit** [comment]
Got a PR (above) out that should fix this.

**Comment by doncarron** [comment]
@MynockSpit We are getting closer. With your PR the key generates closer to being correct but I noticed that it generates with an (extraneous?) space around the variable being passed in:

`Consumer information for <1>{{ name }}</1>`

Instead of

`Consumer information for <1>{{name}}</1>`

Unfortunately the extra space apparently causes the translation to fail to "map" so the string ends up with the fallback language value instead.

**Comment by MynockSpit** [comment]
Oh wow. Yup. Should be a small fix.

**Comment by karellm** [comment]
@MynockSpit Did I merge to PR to quickly or was this fixed? I'll wait before I make a release.

**Comment by MynockSpit** [comment]
I'll have one minor tweak here soon, just as soon as I confirm the behaviour of i18n-react.

**Comment by karellm** [comment]
I just published `1.0.0-beta14` with the fixes
